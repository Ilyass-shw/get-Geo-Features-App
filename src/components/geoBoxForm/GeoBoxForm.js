import React, { useEffect } from "react";
import { useForm } from "react-hook-form";
import { useDispatch, useSelector } from "react-redux";
import GeoboxInput from "../geoboxInput/GeoboxInput";
import {
  getGeoJsonFeatures,
  updateBounds,
} from "../geoJsonFeaturesSlice/geoJsonFeaturesSlice";
import getBounds from "../geoJsonFeaturesSlice/selectors/getBounds";
import getFetchStatus from "../geoJsonFeaturesSlice/selectors/getFetchStatus";
import {
  Form,
  Inputs,
  FirstPoint,
  SecondPoint,
  SubmitBtn,
} from "./GeoBoxFormStyles";

const GeoBoxForm = () => {
  const dispatch = useDispatch();
  const {
    register,
    handleSubmit,
    formState: { errors },
    setValue,
  } = useForm();

  const bounds = useSelector(getBounds);
  const fetchStatus = useSelector(getFetchStatus);
  useEffect(() => {
    setValue("left", bounds.left);
    setValue("right", bounds.right);
    setValue("bottom", bounds.bottom);
    setValue("top", bounds.top);
  }, [setValue, bounds]);
  const onSubmit = (bounds) => {
    dispatch(
      updateBounds({
        left: parseFloat(bounds.left),
        bottom: parseFloat(bounds.bottom),
        right: parseFloat(bounds.right),
        top: parseFloat(bounds.top),
      })
    );
    dispatch(getGeoJsonFeatures());
  };

  return (
    <Form onSubmit={handleSubmit(onSubmit)}>
      <Inputs>
        <FirstPoint>
          <GeoboxInput
            register={register}
            errors={errors}
            name={"left"}
            Min={-180}
            Max={180}
          />

          <GeoboxInput
            register={register}
            errors={errors}
            name={"bottom"}
            Min={-90}
            Max={90}
          />
        </FirstPoint>

        <SecondPoint>
          <GeoboxInput
            register={register}
            errors={errors}
            name={"right"}
            Min={-180}
            Max={180}
          />
          <GeoboxInput
            register={register}
            errors={errors}
            name={"top"}
            Min={-90}
            Max={90}
          />
        </SecondPoint>
      </Inputs>
      <SubmitBtn disabled={fetchStatus === "pending"} type="submit">
        {fetchStatus === "pending" ? "Loading..." : 'Submit'}
      </SubmitBtn>
    </Form>
  );
};

export default GeoBoxForm;
