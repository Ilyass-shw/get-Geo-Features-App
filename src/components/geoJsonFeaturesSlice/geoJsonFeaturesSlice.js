import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";

import getFeaturesFullfiled from "./reducers/getFeaturesFullfiled";
import getFeaturesPending from "./reducers/getFeaturesPending";
import getFeaturesRejected from "./reducers/getFeaturesRejected";

import osmtogeojson from "osmtogeojson";
import updateBoundsReducer from "./reducers/updateBounds";

export const getGeoJsonFeatures = createAsyncThunk(
  "getGeoJsonFeatures",
  async (_, { getState, rejectWithValue }) => {
    const { top, bottom, left, right } = getState().geoJsonFeatures.bounds;
    try {
      const res = await fetch(
        `https://www.openstreetmap.org/api/0.6/map?bbox=${left},${bottom},${right},${top}`
      );
      if (res.ok) {
        // Fetch API doesn't support fetching XML natively, so this is a workaround
        const osmAsString = await res.text();
        const osm = new DOMParser().parseFromString(osmAsString, "text/xml");
        return osmtogeojson(osm);
      } else {
        const failedRes = await res.text();
        return rejectWithValue(failedRes);
      }
    } catch (err) {
      return rejectWithValue(err.message);
    }
  }
);
export const initialState = {
  fetchStatus: "idle",
  error: "",
  geoFeatures: [],
  bounds: {
    left: 13.424494064947247,
    bottom: 52.5328015158109,
    right: 13.426345982999798,
    top: 52.53344512403046,
  },
};

const geoJsonFeaturesSlice = createSlice({
  name: "geoJsonFeatures",
  initialState,
  reducers: {
    updateBounds: updateBoundsReducer,
  },
  extraReducers: (builder) =>
    builder
      .addCase(getGeoJsonFeatures.pending, getFeaturesPending)
      .addCase(getGeoJsonFeatures.rejected, getFeaturesRejected)
      .addCase(getGeoJsonFeatures.fulfilled, getFeaturesFullfiled),
});

export const { updateBounds } = geoJsonFeaturesSlice.actions;

export default geoJsonFeaturesSlice.reducer;
