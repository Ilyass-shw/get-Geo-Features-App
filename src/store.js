import { configureStore } from "@reduxjs/toolkit";
import geoJsonFeaturesSlice from "./components/geoJsonFeaturesSlice/geoJsonFeaturesSlice";

const store = configureStore({
  reducer: {
    geoJsonFeatures: geoJsonFeaturesSlice,
  },
});
export default store;
