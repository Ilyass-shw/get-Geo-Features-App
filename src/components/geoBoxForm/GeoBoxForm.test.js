import store from "../../store";
import { render, screen, waitFor } from "@testing-library/react";
import GeoBoxForm from "./GeoBoxForm";
import { Provider } from "react-redux";
import { updateBounds } from "../geoJsonFeaturesSlice/geoJsonFeaturesSlice";
import userEvent from "@testing-library/user-event";
import { act } from "react-dom/test-utils";

describe("GeoBoxForm", () => {
  it("should render", () => {
    render(
      <Provider store={store}>
        <GeoBoxForm />
      </Provider>
    );
  });
  it("should match snapshot", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <GeoBoxForm />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should have input (bound) values linked to those of store", async () => {
    render(
      <Provider store={store}>
        <GeoBoxForm />
      </Provider>
    );
    const { left, bottom, right, top } =
      store.getState().geoJsonFeatures.bounds;
    expect(screen.getByTestId("left")).toHaveValue(left);
    expect(screen.getByTestId("right")).toHaveValue(right);
    expect(screen.getByTestId("top")).toHaveValue(top);
    expect(screen.getByTestId("bottom")).toHaveValue(bottom);

    store.dispatch(updateBounds({ left: 4, bottom: 9, right: 5, top: 2 }));

    await waitFor(() => {
      expect(screen.getByTestId("left")).toHaveValue(4);
      expect(screen.getByTestId("right")).toHaveValue(5);
      expect(screen.getByTestId("top")).toHaveValue(2);
      expect(screen.getByTestId("bottom")).toHaveValue(9);
    });
  });

  it("should submit correct data to the store", async () => {
    render(
      <Provider store={store}>
        <GeoBoxForm />
      </Provider>
    );

    act(() => {
      userEvent.type(screen.getByTestId("left"), "8");
      userEvent.type(screen.getByTestId("bottom"), "7");
      userEvent.type(screen.getByTestId("right"), "6");
      userEvent.type(screen.getByTestId("top"), "9");
    });

    await waitFor(() => {
      expect(screen.getByTestId("left")).toHaveValue(8);
      expect(screen.getByTestId("right")).toHaveValue(6);
      expect(screen.getByTestId("top")).toHaveValue(9);
      expect(screen.getByTestId("bottom")).toHaveValue(7);
    });

    act(() => {
      userEvent.click(screen.getByRole("button"));
    });

    await waitFor(() => {
      expect(store.getState().geoJsonFeatures.bounds.left).toStrictEqual(8);
      expect(store.getState().geoJsonFeatures.bounds.bottom).toStrictEqual(7);
      expect(store.getState().geoJsonFeatures.bounds.right).toStrictEqual(6);
      expect(store.getState().geoJsonFeatures.bounds.top).toStrictEqual(9);
    });
  });
});
