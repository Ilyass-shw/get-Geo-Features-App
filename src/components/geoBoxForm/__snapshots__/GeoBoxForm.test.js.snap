// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GeoBoxForm should match snapshot 1`] = `
<DocumentFragment>
  <form
    class="sc-eCImPb iuTmwm"
  >
    <div
      class="sc-jRQBWg egnJmN"
    >
      <div
        class="sc-gKclnd"
      >
        <div
          class="sc-bdvvtL ilFTvE"
        >
          <label
            class="sc-dkPtRN hixtpS"
            for="left"
          >
            left
          </label>
          <input
            class="sc-gsDKAQ edMpyo"
            data-testid="left"
            name="left"
            step="any"
            type="number"
          />
        </div>
        <p
          class="sc-hKwDye uccYi"
        />
        <div
          class="sc-bdvvtL ilFTvE"
        >
          <label
            class="sc-dkPtRN hixtpS"
            for="bottom"
          >
            bottom
          </label>
          <input
            class="sc-gsDKAQ edMpyo"
            data-testid="bottom"
            name="bottom"
            step="any"
            type="number"
          />
        </div>
        <p
          class="sc-hKwDye uccYi"
        />
      </div>
      <div
        class="sc-iCfMLu"
      >
        <div
          class="sc-bdvvtL ilFTvE"
        >
          <label
            class="sc-dkPtRN hixtpS"
            for="right"
          >
            right
          </label>
          <input
            class="sc-gsDKAQ edMpyo"
            data-testid="right"
            name="right"
            step="any"
            type="number"
          />
        </div>
        <p
          class="sc-hKwDye uccYi"
        />
        <div
          class="sc-bdvvtL ilFTvE"
        >
          <label
            class="sc-dkPtRN hixtpS"
            for="top"
          >
            top
          </label>
          <input
            class="sc-gsDKAQ edMpyo"
            data-testid="top"
            name="top"
            step="any"
            type="number"
          />
        </div>
        <p
          class="sc-hKwDye uccYi"
        />
      </div>
    </div>
    <button
      class="sc-furwcr TlLRf"
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;
